### PR TITLE
[OpenGL] bitmasked support

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #711 

First problem is bitmasked cannot be assigned correctly:
```py
import taichi as ti

ti.init(arch=ti.opengl)
ti.misc.util.set_gdb_trigger(False)

x = ti.var(ti.i32)
y = ti.var(ti.i32, shape=())
z = ti.var(ti.i32, shape=())
ti.root.bitmasked(ti.i, 4).place(x)

@ti.kernel
def func():
    y[None] = ti.is_active(x.parent(), 0)
    x[0] = 233
    z[None] = ti.is_active(x.parent(), 0)

func()
print(x[0])
print(y[None])
print(z[None])
```
It shows `0 0 0`, it should be at least `233 0 0`.
Good night!

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

**Comment by archibate** [comment]
Metal root buffer layout:
```
SNodeA: i32 | i32 | ...(12)... | i32 | i32 | 00...(12)...00
SNodeB: i32 | i32 | ...(233)... | i32 | i32 | 00...(233)...00
````

OpenGL root buffer layout (for now):
```
Unified bitmask array:  00...(12)...00 | 00...(233)...00
SNodeA: i32 | i32 | ...(12)... | i32 | i32
SNodeB: i32 | i32 | ...(233)... | i32 | i32
```

File: taichi/codegen/codegen_opengl.cpp [hunk 1]
```diff
@@ -233,12 +235,43 @@ class KernelGen : public IRVisitor {
...
+    } else {
+        TI_NOT_IMPLEMENTED;
+    }
+  }
+
+  void visit(GlobalStoreStmt *stmt) override {
+    TI_ASSERT(stmt->width() == 1);
+    auto dt = stmt->data->element_type();
+    auto ch_addr = stmt->ptr->short_name();
+    emit("atomicOr(_rt_.bitmask[{} >> 8], 1 << ({} & 31));", ch_addr, ch_addr);
```
Comment (archibate) [inline taichi/codegen/codegen_opengl.cpp:270]: Is this the correct usage to activate when store?

**Comment by archibate** [comment]
How to struct_for a bitmasked snode? `if (!is_active) return;`? Or something better?

File: taichi/codegen/codegen_opengl.cpp (hunk 1 above)
Comment (yuanming-hu) [reply taichi/codegen/codegen_opengl.cpp:270]: Actually, `activation` should happen on `SNodeLookUp` statements. When doing `GlobalStore` you should assume the element is already activated.

**Comment by yuanming-hu** [comment]
> How to struct_for a bitmasked snode? `if (!is_active) return;`? Or something better?

That depends - for intermediate SNode we need a list generation (see x64/CUDA/Metal backends as examples), for leaf SNode `if (!is_active) return;` is fine.

**Comment by archibate** [comment]
What is a `list generation`?
Why it's fine for leaf SNode like `dense().bitmasked()`?
A non-trival example might be `bitmasked().bitmasked()` or `bitmasked().dense()`?

**Comment by yuanming-hu** [comment]
Good questions - I should probably do my job to improve the documentation #736 .... I plan to do this on Sunday. Sorry for blocking you. If you have extra cycles maybe we can have #653 in. 

Note that it's fine if OpenGL backend in v0.6.0 doesn't support sparse data structures. Metal doesn't support sparse either on its initial release. Most demos so far doesn't require sparsity. 

We should also include glfw and glew as submodules, and statically link their binary into `libtaichi_sore.so` so that we can really ship the OpenGL backend.

**Comment by archibate** [comment]
Right, we'll first ship it, then improve it!
Btw I found #653 too hard thanks to the 30m signal delay from the CI virtual machines...

**Comment by yuanming-hu** [comment]
> Right, we'll first ship it, then improve it!
> Btw I found #653 too hard thanks to the 30m signal delay from the CI virtual machines...

Understood... Setting up CI can be boring due to the long build delay - feel free to do that while watching Bilibili...

**Comment by archibate** [comment]
Won't proceed before doc complete, closing for now.

**Pull request closed without merging**